[VSC-9] Module autocompletions

krendelhoff2 opened the pull request.

krendelhoff2 committed:
[VSC-9] Export some navigation functions

krendelhoff2 committed:
[VSC-9] Populate importable fields cache

krendelhoff2 committed:
[VSC-9] Support module autocompletions

krendelhoff2 committed:
[VSC-9] Fix kind for module names completions

heitor-lassarote commented on src/server/handlers.ts:
```diff
@@ -1171,9 +1171,7 @@ export const addHandlers = (connection: Connection, redirectConsole = true) => {
-                                    kind: path.startsWith('mo:')
-                                        ? CompletionItemKind.Module
-                                        : CompletionItemKind.Class, // TODO: resolve actors, classes, etc.
+                                    kind: CompletionItemKind.Module,
```
Is it still the case if we have the `path` with an `actor` inside? Not sure why it considered `Class` before

krendelhoff2 replied:
Good catch, I think it was `Class` because there is no `actor` completion kind. But I still think `path.startsWith('mo:')` is incorrect way to decide kind, because path could be a library module, not necessarily `actor`.

heitor-lassarote commented on src/server/handlers.ts:
```diff
@@ -1202,39 +1221,52 @@ export const addHandlers = (connection: Connection, redirectConsole = true) => {
+                // Check for an identifier before the dot (e.g. `Module.abc`)
+                const end = position.character - dot.length - identStart.length;
+                const preMatch = /(\s*\.\s*)?([a-zA-Z_][a-zA-Z0-9_]*)$/.exec(
+                    lines[position.line].substring(0, end),
```
One problem with this is multiline expressions, e.g.
```
let example = foo
  .bar
  .baz;
```

heitor-lassarote commented:
Could you add tests? The helpers from the benchmarks can be used for this

krendelhoff2 committed:
[VSC-9] Add completion test

krendelhoff2 committed:
[VSC-9] Add test helper to wait for notification

heitor-lassarote commented on src/server/test/completion.spec.ts:
```diff
@@ -0,0 +1,226 @@
+        server.dispose();
+        client.dispose();
+    });
+    it('local module completion with import 1', async () => {
```
Just a nitpick, but AFAIU, the philosophy of `it` is to write test names in the format "it should ..." or "it is ..." or "it expects ...", etc. So here, for example, it could be `it('provides completions for the "a" module', async () => ...)`.

heitor-lassarote reviewed (commented):
Thank you for the tests, they look good, just one comment.

heitor-lassarote committed:
[VSC-9] Allow completions with multiple lines

Problem: Completions current only work if the entirety of it is written
in a single line.

Solution: Rather than indexing the line at the cursor's line, we instead
match everything between the start of the document and the cursor's
line.

Also write some tests to check that it's working as intended.

krendelhoff2 committed:
fixup! [VSC-9] Support module autocompletions
